### PR TITLE
build: fix warnings for Oracle Developer Studio compilers

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Some compilers, such as Oracle Developer Studio, refuse to work with
empty struct.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
I think I found why `suncc` is so stupid. It treats empty statement as a real statement complains if that empty statement never gets executed. So any pattern like this: 
```
#define M {goto somewhere; }
M;
```
is a warning. Do you see where the empty statement is?

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
Note: as part of the fix, the following `CFLAGS` need be added for `suncc`:
```
-errtags=yes 
-erroff=E_END_OF_LOOP_CODE_NOT_REACHED 
-erroff=E_STATEMENT_NOT_REACHED 
-erroff=E_ANONYMOUS_UNION_DECL 
-erroff=E_EMPTY_TRANSLATION_UNIT
```

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:more/warnings
test:jenkins/ch4/most
test:jenkins/ch3/most

**Comment by hzhou** [comment]
test:more/warnings

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/openpa/src/opa_primitives.h [hunk 1]
```diff
@@ -83,6 +83,8 @@
 #if defined(OPA_USE_UNSAFE_PRIMITIVES)
 /* comes first to permit user overrides in the style of NDEBUG */
 #include "primitives/opa_unsafe.h"
+#elif defined(OPA_HAVE_GCC_INTRINSIC_ATOMICS)
+#include "primitives/opa_gcc_intrinsics.h"
```
Comment (raffenet) [inline src/openpa/src/opa_primitives.h]: Is there any performance data we can look at for this change? I would really like to drop the assembly, too, and my understanding was that moving to MPL atomics would do so. Doing so with OPA and without memory model aware atomics may be premature.

File: src/mpid/ch3/channels/nemesis/include/mpid_nem_datatypes.h [hunk 2]
```diff
@@ -116,12 +116,12 @@
    
 */
 
-#if (SIZEOF_OPA_PTR_T > 8)
```
Comment (raffenet) [inline src/mpid/ch3/channels/nemesis/include/mpid_nem_datatypes.h]: How is this not causing a compile error?

File: src/openpa/src/opa_primitives.h (hunk 1 above)
Comment (hzhou) [reply src/openpa/src/opa_primitives.h]: I probably should remove this commit as it turns out not helping much. Originally I thought gcc intrinsics would compile cleanly, but turns out it emits tons of warning as well. In the end I just turned off both warnings using `#pragma`.

File: src/mpid/ch3/channels/nemesis/include/mpid_nem_datatypes.h (hunk 2 above)
Comment (hzhou) [reply src/mpid/ch3/channels/nemesis/include/mpid_nem_datatypes.h]: Good question. I only know it was set to 0.

**Comment by hzhou** [comment]
test:more/warnings
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
@raffenet I have removed the commit that swaps OPA's primitives preference. The tests seem OK. 2 timeout failures on BSD are believed to be unrelated.

**Pull request merged**